keps/sig-network/NNNN-externalips-admission/README.md

```diff
@@ -0,0 +1,246 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposal is in response to CVE-2020-8554: "Man in the middle using
+LoadBalancer or ExternalIPs".
+
+Fundamentally the `Service.spec.externalIPs[]` feature is bad.  It predates
+`Service.spec.type=LoadBalancer` and, now that we have that, has very few
+use-cases.  In short an unprivileged user can hijack an IP address via a
+Service spec.  In contrast, `type=LoadBalancer` uses Service status, which most
+normal users should not be allowed to write.
+
+This KEP proposes to block the use of ExternalIPs via a built-in admission
+controller.  The justification for this, as opposed to a webhook, is that 99%
+of users will never use this feature, and making them ALL run a webhook seems
+terrible.
+
+## Motivation
+
+https://github.com/kubernetes/kubernetes/issues/97110
+
+### Goals
+
+Make it possible to disable an insecure feature for the vast majority of users
+very quickly.
+
+### Non-Goals
+
+* Make this the default (breaking change)
+* Make the feature safe to use.
+
+## Proposal
+
+This KEP proposes to add a built-in admission controller
+"BlockServiceExternalIPs", which rejects any CREATE or UPDATE operation which
+adds a new value to `Service.spec.externalIPs`.  Existing values will be
+tolerated.
+
+Th number of rejected operations will be exposed as a metric.
+
+### User Stories (Optional)
+
+Alice the admin does not want her users using this insecure feature.  She
+enabled this admission controller and knows no user can use it.  She can then
+audit existing users and make them stop.
+
+### Risks and Mitigations
+
+Some installations may want to use this feature in a more controlled way.  They
+can use a custom webhook admission controller or a policy controller to enforce
+their own rules.
+
+This is a precedent we should not set lightly.  In this case the VAST majority
+of users do not need this feature and it is very surgical in nature.  As far as
+we know, there are no other unprivileged fields with this much power.
+
+## Design Details
+
+One simple admission controller should be enough to disable this misfeature.
+Unfortunately it can not be on by default (that would be breaking).
+
+See "Proposal" above.
+
+### Test Plan
+
+* Unit tests to ensure CREATE and UPDATE operations are rejected when adding
+  new `externalIPs`.
+* Unit tests to ensure UPDATE operations allow existing `externalIPs`.
+
+### Graduation Criteria
+
+#### Alpha -> GA Graduation
+
+- Tests and code complete
+- One cycle in alpha
+
+### Upgrade / Downgrade Strategy
+
+Cluster upgrades/downgrades should not be an issue.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: ExternalIPsAdmissionControl
```

Comment:
Since admission plugins have their own `--enable-admission-plugins`, an additional featuregate seems unnecessary.

Reply:
+1

Reply:
ACK.  I will remove it.